### ISSUE TITLE
"mkfs.xfs -m crc=1" crazes xfs_repair

### ISSUE THREAD

**Issue by akiradeveloper** [akiradeveloper/dm-writeboost#62]
* kernel: 4.1.0-rc2  
* xfsprogs: 3.2.1

formatted a created device (HDD=1G, SSD=2G) by the mkfs.xfs crazes xfs_repair by inconsistent checksum.

```
Phase 1 - find and verify superblock...
Phase 2 - using internal log
        - zero log...
        - scan filesystem freespace and inode maps...
Metadata CRC error detected at block 0x80008/0x1000
btree block 1/1 is suspect, error 74
Metadata CRC error detected at block 0x8/0x1000
Metadata CRC error detected at block 0x80010/0x1000
btree block 1/2 is suspect, error 74
btree block 0/1 is suspect, error 74
Metadata CRC error detected at block 0x180010/0x1000
btree block 3/2 is suspect, error 74
Metadata CRC error detected at block 0x180018/0x1000
btree block 3/3 is suspect, error 74
Metadata CRC error detected at block 0x100008/0x1000
btree block 2/1 is suspect, error 74
Metadata CRC error detected at block 0x100010/0x1000
btree block 2/2 is suspect, error 74
Metadata CRC error detected at block 0x100018/0x1000
btree block 2/3 is suspect, error 74
        - found root inode chunk
```

I am fighting with this bug in highest priority and got some clue that reducing max_io_len from 8 to 4 can fix this problem. But should be working with 8.

**Comment by akiradeveloper** [comment]
Just a debugging note:

The CRC error is detected after mounting the filesystem and try to read the inode metadata. it is sector 24 and the mb->idx is 2780.

```
[   72.458888] XFS (dm-4): Ending clean mount
[   72.515323] device-mapper: writeboost: full read 24 2780
[   72.515598] device-mapper: writeboost: end io 24464
[   72.515732] XFS (dm-4): Metadata CRC error detected at xfs_inobt_read_verify+0xa0/0x230, block 0x18
[   72.516103] XFS (dm-4): Unmount and run xfs_repair
[   72.516338] XFS (dm-4): First 64 bytes of corrupted metadata buffer:
[   72.516612] ffff8802114a1000: 49 41 42 33 00 00 00 01 ff ff ff ff ff ff ff ff  IAB3............
[   72.516953] ffff8802114a1010: 00 00 00 00 00 00 00 18 00 00 00 00 00 00 00 00  ................
[   72.517291] ffff8802114a1020: 8d fc e8 06 a3 9c 4e b4 b4 46 a3 b9 d7 81 3b c8  ......N..F....;.
[   72.517628] ffff8802114a1030: 00 00 00 00 87 dc 72 69 00 00 00 40 00 00 00 3d  ......ri...@...=
[   72.517984] XFS (dm-4): metadata I/O error: block 0x18 ("xfs_trans_read_buf_map") error 74 numblks 8
[   72.620556] XFS (dm-4): Unmounting Filesystem
```

In my script, drop_caches is called just before. And during mounting many overwrite happen.

```
[   71.686640] sh (2917): drop_caches: 3
...
[   71.746881] device-mapper: writeboost: full read 24 2780
[   71.747124] device-mapper: writeboost: end io 24464
...
[   71.791350] XFS (dm-4): Mounting V5 Filesystem
...
[   71.803115] device-mapper: writeboost: delete 2774
[   71.804127] device-mapper: writeboost: delete 195
[   71.805140] device-mapper: writeboost: delete 196
[   71.806335] device-mapper: writeboost: delete 197
[   71.807407] device-mapper: writeboost: delete 198
[   71.808309] device-mapper: writeboost: delete 199
[   71.809355] device-mapper: writeboost: delete 200
[   71.810602] device-mapper: writeboost: delete 201
[   71.811689] device-mapper: writeboost: delete 202
[   71.812810] device-mapper: writeboost: delete 203
[   71.813968] device-mapper: writeboost: delete 204
[   71.815099] device-mapper: writeboost: delete 205
[   71.816730] device-mapper: writeboost: delete 206
[   71.817821] device-mapper: writeboost: delete 207
[   71.818994] device-mapper: writeboost: delete 208
[   71.820037] device-mapper: writeboost: delete 209
[   71.821208] device-mapper: writeboost: delete 210
[   71.830295] device-mapper: writeboost: delete 211
[   71.831589] device-mapper: writeboost: delete 212
[   71.832330] device-mapper: writeboost: delete 213
[   71.833378] device-mapper: writeboost: delete 214
...

```

**Comment by akiradeveloper** [comment]
Printing the partial read reveals that partial read to mb idx 2803 is submitted before full read to 2780.

I suppose the 2803 contains sector 23 and 2780 contains 24-31. 

```
  141.193757] XFS (dm-4): Ending clean mount
[  141.250245] device-mapper: writeboost: part read 2803
[  141.251274] device-mapper: writeboost: full read 24 2780
[  141.251640] device-mapper: writeboost: end io 24464
[  141.251783] XFS (dm-4): Metadata CRC error detected at xfs_inobt_read_verify+0xa0/0x230, block 0x18
[  141.252349] XFS (dm-4): Unmount and run xfs_repair
[  141.252768] XFS (dm-4): First 64 bytes of corrupted metadata buffer:
[  141.253236] ffff8801fccf0000: 49 41 42 33 00 00 00 01 ff ff ff ff ff ff ff ff  IAB3............
[  141.253954] ffff8801fccf0010: 00 00 00 00 00 00 00 18 00 00 00 00 00 00 00 00  ................
[  141.254667] ffff8801fccf0020: b2 d4 c6 54 f2 3e 40 fc 93 04 c9 44 dc fe 22 4a  ...T.>@....D.."J
[  141.255392] ffff8801fccf0030: 00 00 00 00 8b 7f 08 1f 00 00 00 40 00 00 00 3d  ...........@...=
[  141.256111] XFS (dm-4): metadata I/O error: block 0x18 ("xfs_trans_read_buf_map") error 74 numblks 8
[  141.300614] XFS (dm-4): Unmounting Filesystem
```

**Comment by akiradeveloper** [comment]
fixed in v2.0.1
